hazelcast-hibernate5/src/main/java/com/hazelcast/hibernate/VersionAwareMapMergePolicy.java

```diff
@@ -16,45 +16,45 @@
 
 package com.hazelcast.hibernate;
 
-import com.hazelcast.core.EntryView;
-import com.hazelcast.map.merge.MapMergePolicy;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.spi.merge.MergingValue;
+import com.hazelcast.spi.merge.SplitBrainMergePolicy;
 import org.hibernate.cache.spi.entry.CacheEntry;
-
 import java.io.IOException;
-
 /**
  * A merge policy implementation to handle split brain remerges based on the timestamps stored in
  * the values.
  */
-public class VersionAwareMapMergePolicy implements MapMergePolicy {
+public class VersionAwareMapMergePolicy implements SplitBrainMergePolicy<Object, MergingValue<Object>> {
 
-    @Override
-    public Object merge(final String mapName, final EntryView mergingEntry, final EntryView existingEntry) {
-        final Object existingValue = existingEntry != null ? existingEntry.getValue() : null;
-        final Object mergingValue = mergingEntry.getValue();
-        if (existingValue != null && existingValue instanceof CacheEntry
-                && mergingValue != null && mergingValue instanceof CacheEntry) {
 
-            final CacheEntry existingCacheEntry = (CacheEntry) existingValue;
-            final CacheEntry mergingCacheEntry = (CacheEntry) mergingValue;
+    public Object merge(MergingValue<Object> mergingValue, MergingValue<Object> existingValue) {
+
+        final Object mergingVal = mergingValue.getValue();
+        final Object existingVal = existingValue.getValue();
+
+        if (existingVal == null) {
+            return mergingVal;
+        }
+
```

Comment:
adding null check for `mergingVal` and returning `existingVal` if it is `null` would be good. WDYT?

Reply:
Do you mean returning the not null one of the two values if just one of them is null and the other one is not?   

Otherwise, why do we ignore one of the values (null check for existing value or merging value only) and returning the other one?

Reply:
>Otherwise, I could not get the point of checking only one of the values and returning the other one if the former is null.
yes, you are right it is meaningless 👍 
> Do you mean returning the not null one of the two values if just one of them is null and the other one is not?
sounds reasonable so you can keep null checks at current code base.

Btw, could you apply these changes to https://github.com/hazelcast/hazelcast-hibernate repo as well?

Reply:
AFAIK, we are dropping the support for hibernate 3 and 4 in Hazelcast 4.0. If not, I can apply them along with the other incompatible modules used in hazelcast-hibernate 3 & 4 (i.e. UnsafeHelper package etc.)

@mesutcelik what about those supports in Hazelcast 4.0?